File: JS/Counter/protoGen.js
const fs = require('fs')

function main() {
    fs.readFile('Plugin/logger.json', 'utf8', (err, data) => {
        if (err) {
            console.log(`Error reading logger.json: ${err}`)
            return
        }

        let config
        try {
            config = JSON.parse(data)
        } catch (parseErr) {
            console.log(`Error parsing logger.json: ${parseErr}`)
            return
        }

        generateProtobuf(config)
    })
}

function generateProtobuf(config) {
    let protoBuilder = ''

    protoBuilder += `syntax = "${config.syntax}";\n\n`
    protoBuilder += `package ${config.package};\n`
    protoBuilder += `option go_package = "${config.go_package}";\n\n`

    protoBuilder += `message ${config.message.name} {\n`
    let fieldNumber = 1
    for (const [name, fieldType] of Object.entries(config.message.fields)) {
        protoBuilder += `    ${fieldType} ${name} = ${fieldNumber};\n`
        fieldNumber++
    }
    protoBuilder += '}\n'

    const protoFilename = 'Plugin/logger.proto'
    fs.writeFile(protoFilename, protoBuilder, 'utf8', (writeErr) => {
        if (writeErr) {
            console.log(`Error writing protobuf file: ${writeErr}`)
            return
        }

        console.log(`Generated protobuf file: ${protoFilename}`)
    })
}

main()

File: Go/Counter/protoGen.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"strings"
)

type ProtobufField struct {
	Name string `json:"name"`
	Type string `json:"type"`
}

type ProtobufMessage struct {
	Name   string            `json:"name"`
	Fields map[string]string `json:"fields"`
}

type PluginConfig struct {
	Syntax    string          `json:"syntax"`
	Package   string          `json:"package"`
	GoPackage string          `json:"go_package"`
	Message   ProtobufMessage `json:"message"`
}

func main() {

	fileData, err := ioutil.ReadFile("Plugin/logger.json")
	if err != nil {
		fmt.Printf("Error reading logger.json: %v\n", err)
		return
	}

	var config PluginConfig
	if err := json.Unmarshal(fileData, &config); err != nil {
		fmt.Printf("Error parsing logger.json: %v\n", err)
		return
	}

	generateProtobuf(config)
}

func generateProtobuf(config PluginConfig) {
	var protoBuilder strings.Builder

	protoBuilder.WriteString(fmt.Sprintf("syntax = \"%s\";\n\n", config.Syntax))
	protoBuilder.WriteString(fmt.Sprintf("package %s;\n", config.Package))
	protoBuilder.WriteString(fmt.Sprintf("option go_package = \"%s\";\n\n", config.GoPackage))

	protoBuilder.WriteString(fmt.Sprintf("message %s {\n", config.Message.Name))
	fieldNumber := 1
	for name, fieldType := range config.Message.Fields {
		protoBuilder.WriteString(fmt.Sprintf("    %s %s = %d;\n", fieldType, name, fieldNumber))
		fieldNumber++
	}
	protoBuilder.WriteString("}\n")

	protoFilename := "Plugin/logger.proto"
	err := ioutil.WriteFile(protoFilename, []byte(protoBuilder.String()), 0644)
	if err != nil {
		fmt.Printf("Error writing protobuf file: %v\n", err)
		return
	}

	fmt.Printf("Generated protobuf file: %s\n", protoFilename)
}
